Description: Restore budgie window manager animations
 Mutter v7 removed the 'scaled-gravity' effect and reworked
 scaled-gravity effect but rewritten using the now expected
 clutter API calls.
Author: David Mohammed <fossfreedom@ubuntu.com>
Bug: https://github.com/solus-project/budgie-desktop/issues/2040
Forwarded: TBD
Last-Update: 2020-10-15

--- a/src/wm/wm.vala
+++ b/src/wm/wm.vala
@@ -112,14 +112,14 @@
 }
 
 public class MinimizeData {
-    public double scale_x;
-    public double scale_y;
-    public double place_x;
-    public double place_y;
-    public double old_x;
-    public double old_y;
+    public float scale_x;
+    public float scale_y;
+    public float place_x;
+    public float place_y;
+    public float old_x;
+    public float old_y;
 
-    public MinimizeData(double sx, double sy, double px, double py, double ox, double oy) {
+    public MinimizeData(float sx, float sy, float px, float py, float ox, float oy) {
         scale_x = sx;
         scale_y = sy;
         place_x = px;
@@ -820,7 +820,7 @@
         finalize_animations(actor as Meta.WindowActor);
     }
 
-    const int MINIMIZE_TIMEOUT = 195;
+    const int MINIMIZE_TIMEOUT = 225;
 
     public override void minimize(Meta.WindowActor actor)
     {
@@ -851,19 +851,23 @@
         actor.transitions_completed.connect(minimize_done);
 
         /* Save the minimize state for later restoration */
-        var scale_x = (double)(icon.width / actor.width);
-        var scale_y = (double)(icon.height / actor.height);
-        var place_x = (double)icon.x;
-        var place_y = (double)icon.y;
-        var old_x = actor.x;
-        var old_y = actor.y;
+        var scale_factor = Meta.Backend.get_backend ().get_settings ().get_ui_scaling_factor ();
+        var scale_x = (float)((icon.width * scale_factor) / actor.width);
+        var scale_y = (float)((icon.height * scale_factor) / actor.height);
+        var place_x = (float)icon.x * scale_factor;
+        var place_y = (float)icon.y * scale_factor;
+        var old_x = (float)actor.x;
+        var old_y = (float)actor.y;
 
         MinimizeData d = new MinimizeData(scale_x, scale_y, place_x, place_y, old_x, old_y);
 
         actor.set_data("_minimize_data", d);
-        actor.set("opacity", 0U, "scale-gravity", Clutter.Gravity.NORTH_WEST,
-                  "x", d.place_x, "y", d.place_y, "scale-x",
-                  d.scale_x, "scale-y", d.scale_y);
+        actor.set_scale(d.scale_x, d.scale_y);
+        actor.set_x(d.place_x);
+        actor.set_y(d.place_y);
+        actor.opacity = 0U;
+        actor.set_content_gravity(Clutter.ContentGravity.TOP_LEFT);
+        actor.set_pivot_point(0f, 0f);
         actor.restore_easing_state();
     }
 
@@ -876,7 +880,7 @@
         finalize_animations(actor as Meta.WindowActor);
     }
 
-    const int UNMINIMIZE_TIMEOUT = 170;
+    const int UNMINIMIZE_TIMEOUT = 200;
 
     /**
      * Handle unminimize animation
@@ -896,9 +900,11 @@
 
         finalize_animations(actor);
 
-        actor.set("opacity", 0U, "scale-gravity", Clutter.Gravity.NORTH_WEST,
-                  "x", d.place_x, "y", d.place_y, "scale-x",
-                  d.scale_x, "scale-y", d.scale_y);
+        actor.set_pivot_point(0f, 0f);
+        actor.set_scale (d.scale_x, d.scale_y);
+        actor.set_x(d.place_x);
+        actor.set_y(d.place_y);
+        actor.opacity = 0U;
 
         actor.show();
 
@@ -906,8 +912,10 @@
         actor.set_easing_mode(Clutter.AnimationMode.EASE_OUT_QUAD);
         actor.set_easing_duration(UNMINIMIZE_TIMEOUT);
 
-        actor.set("scale-x", 1.0, "scale-y", 1.0, "opacity", 255U,
-                  "x", d.old_x, "y", d.old_y);
+        actor.set_scale (1.0f, 1.0f);
+        actor.opacity = 255U;
+        actor.set_x(d.old_x);
+        actor.set_y(d.old_y);
 
         actor.transitions_completed.connect(unminimize_done);
         state_map.insert(actor, AnimationState.UNMINIMIZE);
@@ -1193,8 +1201,8 @@
     }
 
     /* Return sorted list of user open tabs */
-    private List<weak Meta.Window> get_current_tabs(Meta.Display display, 
-                     Meta.Workspace workspace, 
+    private List<weak Meta.Window> get_current_tabs(Meta.Display display,
+                     Meta.Workspace workspace,
                      bool getTabsForAllWindows)
     {
         List<weak Meta.Window> tabs;
--- a/vapi/libmutter-6.vapi
+++ b/vapi/libmutter-6.vapi
@@ -154,6 +154,7 @@
 		public static unowned Meta.Backend get_backend ();
 		public unowned Meta.Dnd get_dnd ();
 		public unowned Meta.RemoteAccessController get_remote_access_controller ();
+		public unowned Meta.Settings get_settings ();
 		public unowned Clutter.Actor get_stage ();
 		public void lock_layout_group (uint idx);
 		public void set_keymap (string layouts, string variants, string options);
--- a/vapi/libmutter-7.vapi
+++ b/vapi/libmutter-7.vapi
@@ -156,6 +156,7 @@
 		public static unowned Meta.Backend get_backend ();
 		public unowned Meta.Dnd get_dnd ();
 		public unowned Meta.RemoteAccessController get_remote_access_controller ();
+		public unowned Meta.Settings get_settings ();
 		public unowned Clutter.Actor get_stage ();
 		public bool is_rendering_hardware_accelerated ();
 		public void lock_layout_group (uint idx);
